### PR TITLE
feat(IamAssumeAuthenticator): introduce a new authenticator

### PR DESCRIPTION
**Pull request by pyrooka**
This commit introduces the new `IAMAssumeAuthenticator` which will fetch an IAM access token using the IAM `get_token` operation's "assume" grant type. The resulting access token allows the application to assume the identity of a trusted profile, similar to the "sudo" feature of Linux.

### CODE REVIEW

File: test/test_utils.py [hunk 1]
```diff
@@ -286,6 +286,20 @@ def test_get_authenticator_from_credential_file():
     assert authenticator.token_manager.scope is None
     del os.environ['IBM_CREDENTIALS_FILE']
 
+    file_path = os.path.join(os.path.dirname(__file__), '../resources/ibm-credentials-iam-assume.env')
+    os.environ['IBM_CREDENTIALS_FILE'] = file_path
+    authenticator = get_authenticator_from_environment('service 1')
```
Comment (padamstx) [inline test/test_utils.py:291]: "I may not always live on the edge, but when I do... I use spaces in my service names..."  :joy:

**Review by padamstx — APPROVED** [review]
LGTM, although I was able to demonstrate my reviewer's prowess by finding a couple of small typos :joy:

File: Authentication.md [hunk 2]
```diff
@@ -199,6 +200,120 @@ service = ExampleServiceV1.new_instance(service_name='example_service')
 ```
 
 
+## Identity and Access Management (IAM) Authentication (grant type: assume)
+The `IAMAssumeAuthenticator` performs a two-step token fetch sequence to obtain
```
[the remark this answers is not in the document]
Comment (dpopp07) [reply Authentication.md:204]: That is interesting - perhaps capitalizing acronyms is idiomatic in Python?

File: Authentication.md [hunk 3]
```diff
@@ -199,6 +200,120 @@ service = ExampleServiceV1.new_instance(service_name='example_service')
...
+for which an access token should be fetched.
+Exactly one of iam_profile_crn, iam_profile_id or iam_profile_name must be specified.
+
+- iam_profile_name: (optional) the name associated with the trusted profile
+for which an access token should be fetched. When specifying this property, you must also
+specify the iam_account_id property as well.
+Exactly one of iam_profile_crn, iam_profile_id or iam_profile_name must be specified.
+
+- iam_account_id: (optional) the ID associated with the IAM account that contains the trusted profile
+referenced by the iam_profile_name property. The imaAccountId property must be specified if and only if
```
[the remark this answers is not in the document]
Comment (pyrooka) [reply Authentication.md:238]: I have no excuse... 😞

File: Authentication.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (pyrooka) [reply Authentication.md:204]: I have to argue here, because I don't think it's unfortunate. :) I think it's more readable, or at least easier to understand that the part of the word is an acronym - especially when you are working with a unfamiliar code base.

---

bad joke: ON
`IamAuthenticator` --> What? I am not an authenticator...
`IAMAuthenticator` --> Ohh, that must be Identity and Access Management, I got it!
bad joke: OFF

---

> That is interesting - perhaps capitalizing acronyms is idiomatic in Python?

It is! From the official PEP styleguide: `Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.` ([source](https://peps.python.org/pep-0008/#descriptive-naming-styles))

This is exactly the same in Go. I thought it's in the "Effective Go" but I was not able to find it there, so here are a few examples from the stdlib instead: [type DB struct](https://pkg.go.dev/database/sql#DB), [ListenAndServerTLS](https://pkg.go.dev/net/http#ListenAndServeTLS), [ParseHTTPVersion](https://pkg.go.dev/net/http#ParseHTTPVersion)

File: Authentication.md [hunk 4]
```diff
@@ -199,6 +200,120 @@ service = ExampleServiceV1.new_instance(service_name='example_service')
...
+associated with the input apikey. If you specify iam_profile_name, then you must also specify the iam_account_id property
+to indicate the IAM account in which the named trusted profile can be found.
+
+### Programming example
+```python
+from ibm_cloud_sdk_core.authenticators import IAMAssumeAuthenticator
+from <sdk-package-name>.example_service_v1 import *
+
+# Create the authenticator.
+authenticator = IAMAssumeAuthenticator('myapikey')
```
[the remark this answers is not in the document]
Comment (pyrooka) [reply Authentication.md:291]: Oops, a shameless copy-paste error. :)

File: ibm_cloud_sdk_core/authenticators/iam_assume_authenticator.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                'Please remove any surrounding {, }, or \" characters.'
+            )
+
+        # Only one of the following arguments must be specified.
+        mutually_exclusive_attributes = [
+            self.token_manager.iam_profile_id,
+            self.token_manager.iam_profile_crn,
+            self.token_manager.iam_profile_name,
+        ]
+        if list(map(bool, mutually_exclusive_attributes)).count(True) != 1:
```
[the remark this answers is not in the document]
Comment (pyrooka) [reply ibm_cloud_sdk_core/authenticators/iam_assume_authenticator.py:145]: This is the result of being too lazy to type multiple lines, but I'm glad I could inspire you! 😂

File: ibm_cloud_sdk_core/get_authenticator.py [hunk 6]
```diff
@@ -104,6 +106,19 @@ def __construct_authenticator(config: dict) -> Authenticator:
             disable_ssl_verification=config.get('AUTH_DISABLE_SSL', 'false').lower() == 'true',
             scope=config.get('SCOPE'),
         )
+    elif auth_type == Authenticator.AUTHTYPE_IAM_ASSUME.lower():
+        authenticator = IAMAssumeAuthenticator(
+            apikey=config.get('APIKEY'),
```
[the remark this answers is not in the document]
Comment (pyrooka) [reply ibm_cloud_sdk_core/get_authenticator.py:111]: Yes, exactly. It can be specified either way, but placing it before the asterisk without a default value means it's required.

File: test/test_utils.py (hunk 1 above)
Comment (pyrooka) [reply test/test_utils.py:291]: LOL! We have a few more places where we use service names with spaces in it, but those were also me (😂) and a former contributor. I guess this also covers a scenario so basically a test improvement, hah!

File: test/test_iam_assume_token_manager.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+def test_request_token_with_profile_id():
+    iam_url = "https://iam.cloud.ibm.com/identity/token"
+    response = {
+        "access_token": ACCESS_TOKEN,
+        "token_type": "Bearer",
+        "expires_in": 3600,
+        "expiration": _get_current_time() + 3600,
+        "refresh_token": "jy4gl91BQ",
+    }
+    responses.add(responses.POST, url=iam_url, body=json.dumps(response), status=200)
```
[the remark this answers is not in the document]
Comment (pyrooka) [reply test/test_iam_assume_token_manager.py]: No. I mean it's not necessary just for checking the outgoing requests (which the test does) because the mock response is generic enough to cover both HTTP calls and it's not a stack, so matched mock responses are not removed. That said, I changed the reworked these token manager tests a little bit, so now it should mimic a real world scenario, because it returns different results based on the `grant_type`. I also added a test for the `get_token()` method where the returned access token is checked, just to make sure the final result contains the correct one. So please take a look at the new tests, just to make sure I didn't miss anything.:)

File: Authentication.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (padamstx) [reply Authentication.md:204]: ok, so it's the other three languages that are incorrect... my mistake :)

**Review by padamstx — APPROVED** [review]
Still LGTM

File: Authentication.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (pyrooka) [reply Authentication.md:204]: Not necessarily. I think it depends on the language and what the official style guide - or the one that we want to follow - says. Sorry if I was too harsh!

File: test/test_iam_assume_token_manager.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    access_token = token_manager.get_token()
+    # Now the access token should be set along with the profile ID.
+    assert token_manager.request_payload.get('access_token') == ACCESS_TOKEN
+    assert token_manager.request_payload.get('profile_id') == MY_PROFILE_ID
+    assert token_manager.request_payload.get('profile_crn') is None
+    assert token_manager.request_payload.get('profile_name') is None
+    assert token_manager.request_payload.get('account_id') is None
+
+    # The final result should be the other access token, which belong to the "assume" request.
+    assert access_token == OTHER_ACCESS_TOKEN
```
Comment (dpopp07) [inline test/test_iam_assume_token_manager.py:192]: Will you add an assertion that the refresh token is not set?

**Review by dpopp07 — APPROVED** [review]
Looks good! Just want to make sure we're handling the refresh token consistently with Node, and had one other question for my own edification

File: ibm_cloud_sdk_core/authenticators/container_authenticator.py [hunk 9]
```diff
@@ -66,6 +66,7 @@ class ContainerAuthenticator(IAMRequestBasedAuthenticator):
 
     def __init__(
         self,
+        *,
```
[the remark this answers is not in the document]
Comment (pyrooka) [reply ibm_cloud_sdk_core/authenticators/container_authenticator.py:69]: Arguments before the asterisk are positional argument, so they can be passed like `call_function(1, 2)`. Arguments after the `*` become keyword only, so they cannot be used in the previous way, only like this: `call_function(a=1, b=2)`. It's not advised to have too many positional arguments because for example to specify the last one, you have to specify all the previous one too. Since they are used by their position not their keyword.
TLDR; there is a [new `pylint` rule](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-positional-arguments.html) that complains about this.
Yes, this changes the behavior a little, but I don't think this would be a breaking change. If we start getting negative feedback, we can restore the original behavior pretty easily and exclude the affected functions/methods from the linter.

File: test/test_iam_assume_token_manager.py (hunk 8 above)
Comment (pyrooka) [reply test/test_iam_assume_token_manager.py:193]: Good catch. It turned out I only made this change in the authenticator and not in the token manager, so all these properties were accessible: `refresh_token`, `client_id`, `client_secret`

File: ibm_cloud_sdk_core/authenticators/iam_assume_authenticator.py [hunk 10]
```diff
@@ -106,7 +106,7 @@ def __init__(
     def __getattribute__(self, name: str) -> Any:
         disallowed_attrs = ['set_scope', 'set_client_id_and_secret']
         if name in disallowed_attrs:
-            raise AttributeError(f"'IAMAssumeAuthenticator' has no attribute '{name}'")
+            raise AttributeError(f"'{self.__class__.__name__}' has no attribute '{name}'")
```
Comment (pyrooka) [inline ibm_cloud_sdk_core/authenticators/iam_assume_authenticator.py]: Now we won't have to care about changing this if we decide to rename the class. :)

File: test/test_iam_assume_token_manager.py (hunk 8 above)
Comment (dpopp07) [reply test/test_iam_assume_token_manager.py:193]: So is it also the case that the client ID and secret will, if set, be used in the IAM delegate request but _not_ in the assume request?

**Review by padamstx — APPROVED** [review]
Latest changes look good.
